src/partition/src/multi_dim.rs

```diff
@@ -134,6 +151,133 @@ impl MultiDimPartitionRule {
+        self.partition_columns
+            .iter()
+            .map(|col_name| {
+                record_batch
+                    .column_by_name(col_name)
+                    .context(error::UndefinedColumnSnafu { column: col_name })
+                    .and_then(|array| {
+                        Helper::try_into_vector(array).context(error::ConvertToVectorSnafu)
+                    })
+            })
+            .collect::<Result<Vec<_>>>()
+    }
+
+    pub fn split_record_batch_naive(
+        &self,
+        record_batch: &RecordBatch,
+    ) -> Result<HashMap<RegionNumber, BooleanArray>> {
+        let num_rows = record_batch.num_rows();
+
+        let mut result = self
+            .regions
+            .iter()
+            .map(|region| {
+                let mut builder = BooleanBufferBuilder::new(num_rows);
+                builder.append_n(num_rows, false);
+                (*region, builder)
+            })
+            .collect::<HashMap<_, _>>();
+
+        let cols = self.record_batch_to_cols(record_batch)?;
+        let mut current_row = vec![Value::Null; self.partition_columns.len()];
+        for row_idx in 0..num_rows {
+            self.row_at(&cols, row_idx, &mut current_row)?;
+            let current_region = self.find_region(&current_row)?;
+            let region_mask = result
+                .get_mut(&current_region)
+                .unwrap_or_else(|| panic!("Region {} must be initialized", current_region));
+            region_mask.set_bit(row_idx, true);
+        }
+
+        Ok(result
+            .into_iter()
+            .map(|(region, mut mask)| (region, BooleanArray::new(mask.finish(), None)))
+            .collect())
+    }
+
+    pub fn split_record_batch(
+        &self,
+        record_batch: &RecordBatch,
+    ) -> Result<HashMap<RegionNumber, BooleanArray>> {
+        let num_rows = record_batch.num_rows();
+        let physical_exprs = {
+            let cache_read_guard = self.physical_expr_cache.read().unwrap();
+            if let Some((cached_exprs, schema)) = cache_read_guard.as_ref()
+                && schema == record_batch.schema_ref()
+            {
+                cached_exprs.clone()
+            } else {
+                drop(cache_read_guard); // Release the read lock before acquiring write lock
+
+                let schema = record_batch.schema();
+                let new_cache = self
+                    .exprs
+                    .iter()
+                    .map(|e| e.try_as_physical_expr(&schema))
+                    .collect::<Result<Vec<_>>>()?;
+
+                let mut cache_write_guard = self.physical_expr_cache.write().unwrap();
+                cache_write_guard.replace((new_cache.clone(), schema));
+                new_cache
+            }
+        };
+
+        let mut result: HashMap<u32, BooleanArray> = physical_exprs
+            .iter()
+            .zip(self.regions.iter())
```

Comment:
Can we ensure that the order of physical_exprs is exactly the same as the order of regions? We may also check `physical_exprs.len() == self.regions.len()`.